spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,426 @@
+package org.apache.spark.sql.cassandra
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.catalyst.expressions.{Literal, AttributeReference, Expression, Attribute}
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.{ReadConf, ValidRDDType}
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+/**
+ * Table scan type. It allows to create a Relation for given table, keyspace, cluster and user
+ * defined schema. It also allow to pass [[CassandraConnector]], [[ReadConf]] and [[WriteConf]]
+ * explicitly
+ */
+sealed trait ScanType {
+
+  def makeRelation(
+      table: String,
+      keyspace: String,
+      cluster: Option[String],
+      userSpecifiedSchema: Option[StructType],
+      connector: CassandraConnector,
+      readConf: ReadConf,
+      writeConf: WriteConf,
+      sqlContext: SQLContext) : BaseRelationImpl
+}
+
+/** Base table scan type */
+case object BaseScanType extends ScanType  {
+
+  override def makeRelation(
+               table: String,
+               keyspace: String,
+               cluster: Option[String],
+               userSpecifiedSchema: Option[StructType],
+               connector: CassandraConnector, readConf: ReadConf, writeConf: WriteConf,
+               sqlContext: SQLContext) : BaseRelationImpl = {
+
+    new TableScanRelationImpl(
+        table,
+        keyspace,
+        cluster,
+        connector,
+        readConf,
+        writeConf,
+        userSpecifiedSchema,
+        sqlContext)
+  }
+
+}
+
+/** Pruned columns table scan type */
+case object PrunedScanType extends ScanType {
+
+  override def makeRelation(
+               table: String,
+               keyspace: String,
+               cluster: Option[String],
+               userSpecifiedSchema: Option[StructType],
+               connector: CassandraConnector,
+               readConf: ReadConf,
+               writeConf: WriteConf,
+               sqlContext: SQLContext) : BaseRelationImpl = {
+
+    new PrunedScanRelationImpl(
+        table,
+        keyspace,
+        cluster,
+        connector,
+        readConf,
+        writeConf,
+        userSpecifiedSchema,
+        sqlContext)
+  }
+
+}
+
+/** Pruned columns and filtered table scan type */
+case object PrunedFilteredScanType extends ScanType {
+
+  override def makeRelation(
+               table: String,
```

Comment:
Better now, but there is no reason to align it to def. Stick to 4-spaces indent everywhere (and 2-spaces for `implicit`, `extends` and `with` keywords to make them stick out to the left).